### PR TITLE
Add multiple writer threads to improve throughput

### PR DESCRIPTION
**Pull request by rajubairishetti**
Refer https://jira.corp.inmobi.com/browse/SCRIBE-37

### CODE REVIEW

File: src/scribe_server.cpp [hunk 1]
```diff
@@ -388,38 +389,39 @@ bool scribeHandler::throttleRequest(const vector<LogEntry>&  messages) {
...
+    } else {
+      unsigned long long size = (*store_iter)->getSize();
+      if (size <= maxQueueSize) {
+        throttle = false;
       }
     }
   }
-
+  if (throttle) {
+    incCounter(category, "denied for queue size");
```
Comment (rajubairishetti) [inline src/scribe_server.cpp]: Will log the information of throttle as part of [SCRIBE-30](https://jira.corp.inmobi.com/browse/SCRIBE-30)

File: src/conf.cpp [hunk 2]
```diff
@@ -119,7 +119,8 @@ bool StoreConf::getString(const string& stringName,
   if (storeType.empty()
       || stringName == "type"
       || stringName == "category"
-      || stringName == "categories") {
+      || stringName == "categories"
+      || stringName == "num_store_threads") {
```
Comment (satish-mittal) [inline src/conf.cpp:123]: why is this needed?

File: src/scribe_server.cpp [hunk 3]
```diff
@@ -388,38 +389,39 @@ bool scribeHandler::throttleRequest(const vector<LogEntry>&  messages) {
...
+  }
+  bool throttle = true;
+  for (store_list_t::iterator store_iter = pstores->begin();
+       store_iter != pstores->end(); ++store_iter) {
+    if (*store_iter == NULL) {
+      throw std::logic_error("throttle check: iterator in store map holds null pointer");
+    } else {
+      unsigned long long size = (*store_iter)->getSize();
+      if (size <= maxQueueSize) {
+        throttle = false;
```
Comment (satish-mittal) [inline src/scribe_server.cpp]: Scribe changes to improve throttling logic is tracked in a separate jira SCRIBE-29. Can we undo the throttling change from this jira and have it together in a separate pull request for SCRIBE-29?

File: src/conf.cpp (hunk 2 above)
Comment (rajubairishetti) [inline src/conf.cpp:123]: This is required. Otherwise this property will be inherited to the stores (if we do not specify the num_store_threads conf) which are configured below this store-category in the scribe conf.

File: src/scribe_server.cpp (hunk 3 above)
Comment (rajubairishetti) [inline src/scribe_server.cpp]: yeah..sure.. I will revert this commit

File: src/scribe_server.cpp [hunk 4]
```diff
@@ -439,7 +441,12 @@ shared_ptr<store_list_t> scribeHandler::createNewCategory(
       shared_ptr<store_list_t> pstores = cat_prefix_iter->second;
       for (store_list_t::iterator store_iter = pstores->begin();
           store_iter != pstores->end(); ++store_iter) {
-        createCategoryFromModel(category, *store_iter);
+        // currently we are not supporting multiple threads for the model categories.
+        // specify new thread name for each new store queue
```
Comment (satish-mittal) [inline src/scribe_server.cpp]: modify comments to be more clear.

File: src/scribe_server.cpp [hunk 5]
```diff
@@ -439,7 +441,12 @@ shared_ptr<store_list_t> scribeHandler::createNewCategory(
...
+        ostringstream ostr;
+        ostr << "";
+        std::string thread_name = ostr.str();
```
Comment (satish-mittal) [inline src/scribe_server.cpp]: you may want to instead define a static string empty_str as a global variable and use it.

File: src/scribe_server.cpp [hunk 6]
```diff
@@ -460,7 +467,12 @@ shared_ptr<store_list_t> scribeHandler::createNewCategory(
   if (store_list == NULL && !defaultStores.empty()) {
     for (store_list_t::iterator store_iter = defaultStores.begin();
         store_iter != defaultStores.end(); ++store_iter) {
-      createCategoryFromModel(category, *store_iter);
+      // currently we are not supporting multiple threads for the model categories.
+      // Specify a new threadName for each storeQueue thread in case of multi-thread
```
Comment (satish-mittal) [inline src/scribe_server.cpp]: same as above.

File: src/scribe_server.cpp [hunk 7]
```diff
@@ -500,6 +512,41 @@ void scribeHandler::addMessage(
   }
 }
 
+// Add this message to every least sized store in list
+void scribeHandler::addMessageToLeastSizedQueue(
```
Comment (satish-mittal) [inline src/scribe_server.cpp]: Instead of exposing a new method, can we merge this logic with existing addMessage()?

File: src/scribe_server.cpp [hunk 8]
```diff
@@ -500,6 +512,41 @@ void scribeHandler::addMessage(
...
+
+  for (store_list_t::iterator store_iter = store_list->begin();
+	  store_iter != store_list->end(); ++store_iter) {
+	++numstores;
+  }
+
+  if (numstores) {
+	// find the least sized store queue add the
+	int minIndex = 0;
+	size_t min_queue_size = (*store_list)[0]->getSize();
```
Comment (satish-mittal) [inline src/scribe_server.cpp]: The current scribe code iterates through std::vector using iterator mechanism. E.g. store_list_t::iterator store_iter...

Let us continue with that style for having seamless code and behavior, instead of mixing it with index based iteration, e.g. (*store_list)[i].

File: src/scribe_server.cpp [hunk 9]
```diff
@@ -873,14 +923,35 @@ bool scribeHandler::configureStore(pStoreConf store_conf, int *numstores) {
...
+    } else {
+      for (std::size_t i = 0; i < num_store_threads; i++) {
+        ostringstream ostr;
+        ostr << "thread_" << i;
+        std::string thread_name = ostr.str();
+        LOG_OPER("Store Queue name [%s] for [%s] category", thread_name.c_str(), category_list[0].c_str());
+        shared_ptr<StoreQueue> result =
+            configureStoreCategory(store_conf, category_list[0], model, thread_name);
+        if (result == NULL) {
+          LOG_OPER("Unable to create store queue [%s] for [%s] catgeory", thread_name.c_str(), category_list[0].c_str());
```
Comment (satish-mittal) [inline src/scribe_server.cpp]: fix typo for 'category'

File: src/scribe_server.cpp [hunk 10]
```diff
@@ -927,6 +1002,7 @@ shared_ptr<StoreQueue> scribeHandler::configureStoreCategory(
   pStoreConf store_conf,                       //configuration for store
   const string &category,                      //category name
   const boost::shared_ptr<StoreQueue> &model,  //model to use (optional)
+  string& thread_name,                         //storeQueue name
```
Comment (satish-mittal) [inline src/scribe_server.cpp]: Typically new arguments are added to the end of an existing signature (i.e. after category_list). Also declare it const string&.

File: src/scribe_server.cpp (hunk 5 above)
Comment (satish-mittal) [inline src/scribe_server.cpp]: Actually you can define thread_name as an optional parameter in scribe_server.h, such that it has a default value of "" in function declaration in .h file itself. That way you won't need to pass on default value of "" at each place.

File: src/scribe_server.cpp [hunk 11]
```diff
@@ -873,14 +923,35 @@ bool scribeHandler::configureStore(pStoreConf store_conf, int *numstores) {
...
-
-    if (result == NULL) {
-      return false;
+    bool is_prefix_category = (!category.empty() &&
+			category[category.size() - 1] == '*');
+    bool is_default_category = (!category.empty() && category.compare("default") == 0);
+
+    unsigned long int num_store_threads = -1;
+    if (!store_conf->getUnsigned("num_store_threads", num_store_threads)
+      || is_default_category || is_prefix_category || (num_store_threads <= 0)) {
```
Comment (satish-mittal) [inline src/scribe_server.cpp]: check for num_store_threads <= 1

File: src/scribe_server.h [hunk 12]
```diff
@@ -130,11 +130,13 @@ class scribeHandler : virtual public scribe::thrift::scribeIf,
   void deleteCategoryMap(category_map_t& cats);
   const char* statusAsString(facebook::fb303::fb_status new_status);
   bool createCategoryFromModel(const std::string &category,
-                               const boost::shared_ptr<StoreQueue> &model);
+                               const boost::shared_ptr<StoreQueue> &model,
+                               std::string& thread_name);
```
Comment (satish-mittal) [inline src/scribe_server.h]: make it const and default value as empty str

File: src/store.cpp [hunk 13]
```diff
@@ -81,34 +81,34 @@ bool shouldSendDummy(boost::shared_ptr<logentry_vector_t> messages) {
...
   if (0 == type.compare("file")) {
-    return shared_ptr<Store>(new FileStore(storeq, category, multi_category,
+    return shared_ptr<Store>(new FileStore(storeq, category, thread_name, multi_category,
                                           readable));
   } else if (0 == type.compare("buffer")) {
-    return shared_ptr<Store>(new BufferStore(storeq,category, multi_category));
+    return shared_ptr<Store>(new BufferStore(storeq,category, thread_name, multi_category));
   } else if (0 == type.compare("network")) {
-    return shared_ptr<Store>(new NetworkStore(storeq, category,
+    return shared_ptr<Store>(new NetworkStore(storeq, category, thread_name,
```
Comment (satish-mittal) [inline src/store.cpp]: Currently we are adding this feature for only file store. Hence we should pass thread_name arg only for file and buffer.

File: src/store.cpp [hunk 14]
```diff
@@ -117,9 +117,11 @@ Store::createStore(StoreQueue* storeq, const string& type,
 
 Store::Store(StoreQueue* storeq,
              const string& category,
+             string& thread_name,
```
Comment (satish-mittal) [inline src/store.cpp]: declare it const

File: src/store.cpp [hunk 15]
```diff
@@ -250,6 +252,9 @@ void FileStoreBase::configure(pStoreConf configuration, pStoreConf parent) {
     filePath += "/" + subDirectory;
   }
 
+  if (!isPrimary &&  !threadName.empty() && subDirectory.empty()) {
```
Comment (satish-mittal) [inline src/store.cpp]: We can remove the check for subDirectory.empty since we are not using it to build spooling path.

File: src/store.cpp [hunk 16]
```diff
@@ -622,6 +631,10 @@ void FileStoreBase::setHostNameSubDir() {
   } else {
     subDirectory = hoststring;
   }
+  // append threadName to the subdirectory
+  if (!threadName.empty()) {
+    subDirectory = hoststring + "_" +threadName;
```
Comment (satish-mittal) [inline src/store.cpp]: missing space after +

File: src/store.cpp [hunk 17]
```diff
@@ -750,6 +763,10 @@ bool FileStore::openInternal(bool incrementFilename, struct tm* current_time) {
       success = writeFile->createDirectory(filePath);
     }
 
+    if (!isPrimary && !threadName.empty()) {
```
Comment (satish-mittal) [inline src/store.cpp]: check for success as the first condition inside if

File: src/store.h [hunk 18]
```diff
@@ -255,11 +257,11 @@ class FileStore : public FileStoreBase {
 class ThriftFileStore : public FileStoreBase {
  public:
   ThriftFileStore(StoreQueue* storeq,
-                  const std::string& category,
+                  const std::string& category, std::string& thread_name,
```
Comment (satish-mittal) [inline src/store.h]: Undo all the signature changes to add thread_name in unwanted stores.

File: src/store.cpp [hunk 19]
```diff
@@ -1480,12 +1499,12 @@ void BufferStore::configure(pStoreConf configuration, pStoreConf parent) {
       string msg("Bad config - buffer primary store cannot be multistore");
       setStatus(msg);
     } else {
-      primaryStore = createStore(storeQueue, type, categoryHandled, false,
+      primaryStore = createStore(storeQueue, type, categoryHandled, threadName, false,
                                   multiCategory);
-      primaryStore->configure(primary_store_conf, storeConf);
```
Comment (satish-mittal) [inline src/store.cpp:1485]: do we need the change the order of these 2 calls?

File: src/scribe_server.cpp [hunk 20]
```diff
@@ -439,6 +439,7 @@ shared_ptr<store_list_t> scribeHandler::createNewCategory(
       shared_ptr<store_list_t> pstores = cat_prefix_iter->second;
       for (store_list_t::iterator store_iter = pstores->begin();
           store_iter != pstores->end(); ++store_iter) {
+        // currently we are not supporting multiple threads for prefix model
```
Comment (satish-mittal) [inline src/scribe_server.cpp]: Can you enhance it to: "Currently we are not supporting multiple threads for prefix model. Hence not passing any thread name".

File: src/scribe_server.cpp [hunk 21]
```diff
@@ -481,19 +482,46 @@ void scribeHandler::addMessage(
...
+  size_t min_queue_size = -1;
+  shared_ptr<StoreQueue> min_store_queue;
+  bool isMultiThreaded = false;
+  for (store_list_t::iterator store_iter = store_list->begin();
+         store_iter != store_list->end();
+         ++store_iter) {
+    if (!(*store_iter)->getThreadName().empty()) {
+      isMultiThreaded = true;
+    }
+    break;
```
Comment (satish-mittal) [inline src/scribe_server.cpp:491]: you should move this break to inside the if condition; else you immediately break out of for loop after one iteration.

File: src/scribe_server.cpp [hunk 22]
```diff
@@ -481,19 +482,46 @@ void scribeHandler::addMessage(
...
+  }
+
+  // Find least sized queue in case if category store configured with multiple threads.
```
Comment (satish-mittal) [inline src/scribe_server.cpp]: Rephrase the comment to: 
If category store is configured with multiple threads, then add message to the least sized queue. Else simply add it to all stores in the queue.

File: src/scribe_server.cpp [hunk 23]
```diff
@@ -481,19 +482,46 @@ void scribeHandler::addMessage(
...
+      // add message to every store queue
+      boost::shared_ptr<LogEntry> ptr(new LogEntry);
+      ptr->category = entry.category;
+      ptr->message = entry.message;
 
-    (*store_iter)->addMessage(ptr);
+      (*store_iter)->addMessage(ptr);
+    } else {
+	  if (min_queue_size > (*store_iter)->getSize()) {
+        min_queue_size = (*store_iter)->getSize();
```
Comment (satish-mittal) [inline src/scribe_server.cpp:509]: Fix indentation of these 2 lines inside if condition

File: src/scribe_server.cpp (hunk 21 above)
Comment (rajubairishetti) [inline src/scribe_server.cpp:491]: Here, we just wanted to know whether the category store is configured with multiple threads or not. It should come out from the loop if thread name is not defined for store queue thread.

File: src/scribe_server.cpp [hunk 24]
```diff
@@ -481,19 +484,46 @@ void scribeHandler::addMessage(
...
+    if (!isMultiThreaded) {
+      // add message to every store queue
+      boost::shared_ptr<LogEntry> ptr(new LogEntry);
+      ptr->category = entry.category;
+      ptr->message = entry.message;
 
-    (*store_iter)->addMessage(ptr);
+      (*store_iter)->addMessage(ptr);
+    } else {
+      if (min_queue_size > (*store_iter)->getSize()) {
```
Comment (satish-mittal) [inline src/scribe_server.cpp]: As we discussed, please re-organize the code. Also since min_queue_size is -1 initially, we will never enter this if condition.

File: src/scribe_server.h [hunk 25]
```diff
@@ -143,6 +145,8 @@ class scribeHandler : virtual public scribe::thrift::scribeIf,
     createNewCategory(const std::string& category);
   void addMessage(const scribe::thrift::LogEntry& entry,
                   const boost::shared_ptr<store_list_t>& store_list);
+  void addMessageToLeastSizedQueue(const scribe::thrift::LogEntry& entry,
```
Comment (satish-mittal) [inline src/scribe_server.h]: Why is this method still there?

File: src/store.h [hunk 26]
```diff
@@ -391,10 +399,11 @@ class NetworkStore : public Store {
  public:
   NetworkStore(StoreQueue* storeq,
                const std::string& category,
-               bool multi_category);
+               bool multi_category, const std::string& thread_name=empty_string);
```
Comment (satish-mittal) [inline src/store.h]: why do we need thread name in this ctor?

File: src/scribe_server.cpp [hunk 27]
```diff
@@ -330,7 +330,8 @@ const char* scribeHandler::statusAsString(fb_status status) {
 
 // Should be called while holding a writeLock on scribeHandlerLock
 bool scribeHandler::createCategoryFromModel(
-  const string &category, const boost::shared_ptr<StoreQueue> &model) {
+  const string &category, const boost::shared_ptr<StoreQueue> &model,
+  const string&  thread_name) {
```
Comment (satish-mittal) [inline src/scribe_server.cpp]: Since we have decided that model categories will not support multiple threads as of now, we don't need this signature change.

File: src/scribe_server.cpp [hunk 28]
```diff
@@ -439,6 +439,8 @@ shared_ptr<store_list_t> scribeHandler::createNewCategory(
       shared_ptr<store_list_t> pstores = cat_prefix_iter->second;
       for (store_list_t::iterator store_iter = pstores->begin();
           store_iter != pstores->end(); ++store_iter) {
+        // Currently we are not supporting multiple threads for prefix model.
+        // Hence not passing any thread name
```
Comment (satish-mittal) [inline src/scribe_server.cpp]: Remove this comment since not needed.

File: src/scribe_server.cpp [hunk 29]
```diff
@@ -455,11 +457,12 @@ shared_ptr<store_list_t> scribeHandler::createNewCategory(
     cat_prefix_iter++;
   }
 
-
   // Then try creating a store if we have a default store defined
   if (store_list == NULL && !defaultStores.empty()) {
     for (store_list_t::iterator store_iter = defaultStores.begin();
         store_iter != defaultStores.end(); ++store_iter) {
+      // currently we are not supporting multiple threads for default model.
+      // Hence not passing any thread name
```
Comment (satish-mittal) [inline src/scribe_server.cpp]: Remove this comment.

File: src/scribe_server.cpp [hunk 30]
```diff
@@ -873,14 +903,32 @@ bool scribeHandler::configureStore(pStoreConf store_conf, int *numstores) {
...
+    unsigned long int num_store_threads = -1;
+    if (!store_conf->getUnsigned("num_store_threads", num_store_threads)
+      || is_default_category || is_prefix_category || (num_store_threads <= 1)) {
+      shared_ptr<StoreQueue> result =
+          configureStoreCategory(store_conf, category_list[0], model);
+      if (result == NULL) {
+        return false;
+      }
+    } else {
+      for (std::size_t i = 0; i < num_store_threads; i++) {
```
Comment (satish-mittal) [inline src/scribe_server.cpp:918]: Log a message that prints the value of num_store_threads config.

File: src/scribe_server.cpp [hunk 31]
```diff
@@ -873,14 +903,32 @@ bool scribeHandler::configureStore(pStoreConf store_conf, int *numstores) {
...
+        ostringstream ostr;
+        ostr << "thread_" << i;
+        const std::string thread_name = ostr.str();
+        LOG_OPER("Store Queue name [%s] for [%s] category", thread_name.c_str(), category_list[0].c_str());
```
Comment (satish-mittal) [inline src/scribe_server.cpp]: Move this log message down to the condition when store queue is configured.

File: src/scribe_server.cpp [hunk 32]
```diff
@@ -962,7 +1010,7 @@ shared_ptr<StoreQueue> scribeHandler::configureStoreCategory(
     if (model != NULL) {
       // Create a copy of the model if we want a new thread per category
       if (newThreadPerCategory && !is_default && !is_prefix_category) {
-        pstore = shared_ptr<StoreQueue>(new StoreQueue(model, category));
+        pstore = shared_ptr<StoreQueue>(new StoreQueue(model, category, thread_name));
```
Comment (satish-mittal) [inline src/scribe_server.cpp]: Remove this code since it is the case of model based store creation.

File: src/store.h [hunk 33]
```diff
@@ -55,13 +54,16 @@ class Store {
...
     createStore(StoreQueue* storeq,
                 const std::string& type, const std::string& category,
-                bool readable = false, bool multi_category = false);
+                bool readable = false, bool multi_category = false,
+                const std::string& thread_name = empty_string);
 
   Store(StoreQueue* storeq, const std::string& category,
-        const std::string &type, bool multi_category = false);
+        const std::string &type, bool multi_category = false,
+        const std::string& thread_name = empty_string);
```
Comment (satish-mittal) [inline src/store.h:62]: we don't need to have thread_name in base class if we can keep thread_name inside buffer store itself.

File: src/store.h [hunk 34]
```diff
@@ -55,13 +54,16 @@ class Store {
...
   virtual ~Store();
 
-  virtual boost::shared_ptr<Store> copy(const std::string &category) = 0;
+  virtual boost::shared_ptr<Store> copy(const std::string &category,
+		  const std::string& thread_name = empty_string) = 0;
```
Comment (satish-mittal) [inline src/store.h]: This copy is not needed since this method is called only while creating store queue from a model.

File: src/scribe_server.cpp [hunk 35]
```diff
@@ -873,14 +900,34 @@ bool scribeHandler::configureStore(pStoreConf store_conf, int *numstores) {
...
+    unsigned long int num_store_threads = -1;
+    if (!store_conf->getUnsigned("num_store_threads", num_store_threads)
+      || is_default_category || is_prefix_category || (num_store_threads <= 1)) {
+      shared_ptr<StoreQueue> result =
+        configureStoreCategory(store_conf, category_list[0], model);
+      if (result == NULL) {
+        return false;
+      }
+    } else {
+      LOG_OPER("Configuring [%lu] store queues for [%s] ", num_store_threads, category_list[0].c_str());
```
Comment (satish-mittal) [inline src/scribe_server.cpp]: category_list[0].c_str() is being repeated multiple times. You could keep a local var for it.

File: src/store.cpp [hunk 36]
```diff
@@ -644,8 +651,9 @@ void FileStoreBase::auditFileClosed() {
 
 FileStore::FileStore(StoreQueue* storeq,
                      const string& category,
-                     bool multi_category, bool is_buffer_file)
-  : FileStoreBase(storeq, category, "file", multi_category),
+                     bool multi_category, bool is_buffer_file,
+                     const string &thread_name)
+  : FileStoreBase(storeq, category, "file", multi_category, thread_name),
```
Comment (satish-mittal) [inline src/store.cpp:656]: In addition to stream level counters, can we introduce a stream_thread level counter and increment it? That will help us get more insights into the write performance at thread level.

**Comment by satish-mittal** [comment]
You need to run benchmark at local/merge level to ensure that these scribe changes work fine with current conduit worker (i.e. no changes required in conduit worker).

**Comment by satish-mittal** [comment]
+1
Please run benchmark tests for merge also.
